[RISCV] Support RISCV Atomics ABI attributes

ilovepi opened the pull request:
This patch adds support for the `atomic_abi` attribute, specifid in
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version.

The atomics_abi tag merging is conducted as follows:

- UNKNOWN is safe to merge with all other values.
- A6C is compatible with A6S, and results in the A6C ABI.
- A6C is incompatible with A7, and results in an error.
- A6S and A7 are compatible, and merging results in the A7 ABI.

Note: the A7 is not yet supported in either LLVM or in any current hardware,
and is therefore ommited from attribute generation in RISCVTargetStreamer.

ilovepi committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

ilovepi committed:
Fix formatting

Created using spr 1.3.4

petrhosek commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -1084,10 +1084,86 @@ static void mergeArch(RISCVISAInfo::OrderedExtensionMap &mergedExts,
+                        const InputSectionBase *newSection, unsigned int oldTag,
+                        unsigned int newTag) {
+  using RISCVAttrs::RISCVAtomicAbiTag::AtomicABI;
+  llvm::errs() << "oldTag=" << oldTag << ", newTag=" << newTag << "\n";
```
Leftover debug print?

ilovepi replied:
Oh I missed that. Thanks.

ilovepi replied:
done.

petrhosek commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -1084,10 +1084,86 @@ static void mergeArch(RISCVISAInfo::OrderedExtensionMap &mergedExts,
+  };
+}
+
+static void mergeX3RegUse(DenseMap<unsigned, unsigned> &intAttr,
```
Where is this function used?

ilovepi replied:
Shoot this belongs in another PR. I’ll update these in the morning. Thanks for pointing that out.

ilovepi committed:
Remove code from SCS PR

Created using spr 1.3.4

ilovepi committed:
remove debugging statements

Created using spr 1.3.4

ilovepi committed:
Remove unrelated code

Created using spr 1.3.4

topperc commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -1084,10 +1084,63 @@ static void mergeArch(RISCVISAInfo::OrderedExtensionMap &mergedExts,
+  auto attr = RISCVAttrs::ATOMIC_ABI;
+  switch (oldTag) {
+  case AtomicABI::UNKNOWN:
+    intAttr[attr] = newTag;
```
Can we use the original iterator returned by try_emplace to avoid the second map lookup here?

ilovepi replied:
done.

ilovepi committed:
Pass iterator instead of repeating lookup

Created using spr 1.3.4

ilovepi committed:
Fix whitespace

Created using spr 1.3.4

ilovepi committed:
Update tests and fix whitespace

Created using spr 1.3.4

ilovepi commented:
@jrtc27 Do you have thoughts on this change w.r.t. the defaults? I recall you had expressed concerns previously in the psABI, so I want to be sure we're changing the status quo appropriately.

jrtc27 commented:
> @jrtc27 Do you have thoughts on this change w.r.t. the defaults? I recall you had expressed concerns previously in the psABI, so I want to be sure we're changing the status quo appropriately.

To be clear, without reading the patch in detail, this is making the default be A6S? I don't think I have concerns with that other than to note that transitioning from A6S to A7 will be an ABI break for existing binaries, so the timeline for when that's ok to do is unclear.

ilovepi committed:
Add NEXT to CHECK lines in test

Created using spr 1.3.4

ilovepi commented:
ping. I think everything has been addressed.

asb commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:
```diff
@@ -75,6 +75,13 @@ void RISCVTargetStreamer::emitTargetAttributes(const MCSubtargetInfo &STI,
+
+  if (STI.hasFeature(RISCV::FeatureStdExtA)) {
+    unsigned AtomicABITag = STI.hasFeature(RISCV::FeatureTrailingSeqCstFence)
+                                ? RISCVAttrs::RISCVAtomicAbiTag::AtomicABI::A6S
```
Is this the right way round? Isn't `STI.hasFeature(RISCV::FeatureTrailingSeqCstFence)` true if `+no-seq-cst-trailing-fence` is present? And if there's no trailing fence, then we have A6C? Or am I misreading?

I think it would be better to name the feature `FeatureNoTrailingSeqCstFence` as we do for `FeatureNoRVCHints`. Though I think this patch would be best if it aimed to change no defaults at all, and left that as a follow-on patch (which would hopefully be a quick LGTM as I think there's no objection to that, going by this review thread).

asb replied:
> I think this should be correct. The default is now "enabled", so adding `+no-seq-cst-trailing-fence` should make `STI.hasFeature(RISCV::FeatureTrailingSeqCstFence)` return `false`. I think the tests indicate that's how its working.

So in attributes.ll, the A6S output has the RUN line `; RUN: llc -mtriple=riscv64 -mattr=+a,+no-seq-cst-trailing-fence %s -o - | FileCheck --check-prefixes=CHECK,RV64A,A6S %s`. But for A6S, don't we require that the trailing fence is emitted?

ilovepi replied:
huh, you're right. Let me double check that now that the default is in it's own change.

ilovepi committed:
Split out change of defaults into separate patch

Created using spr 1.3.4

asb commented:
Thanks for splitting out that change. I'm thinking you might be missing a test case in lld/test/ELF/riscv-attributes.s. You handle an input file where the atomic_abi attribute is present and explicitly set to `UNKNOWN`. But what about when merging a file where the attribute was explicitly set, with one where the attribute wasn't present at all?

ilovepi commented:
> Thanks for splitting out that change. I'm thinking you might be missing a test case in lld/test/ELF/riscv-attributes.s. You handle an input file where the atomic_abi attribute is present and explicitly set to `UNKNOWN`. But what about when merging a file where the attribute was explicitly set, with one where the attribute wasn't present at all?

Oh, that's a good catch. I didn't think about that case. Let me see what happens.

ilovepi committed:
Improve testing

Created using spr 1.3.4

asb commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -62,12 +62,21 @@
 # RUN: llvm-readobj -A atomic_abi_A6C_unknown | FileCheck %s --check-prefixes=UNKNOWN_A6C
 
+# RUN: ld.lld atomic_abi_unknown.o diff_stack_align.o -o atomic_abi_none_unknown
+# RUN: llvm-readobj -A atomic_abi_none_unknown | FileCheck %s --check-prefixes=UNKNOWN_UNKNOWN
```
Shouldn't the check-prefix for this be NONE_UNKNOWN to match the convention elsewhere?

asb commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -62,12 +62,21 @@
+# RUN: ld.lld atomic_abi_unknown.o diff_stack_align.o -o atomic_abi_none_unknown
+# RUN: llvm-readobj -A atomic_abi_none_unknown | FileCheck %s --check-prefixes=UNKNOWN_UNKNOWN
+
+# RUN: ld.lld diff_stack_align.o atomic_abi_A6C.o -o atomic_abi_A6C_unknown
```
Shouldn't the output filename be atomic_abi_A6C_none?

asb commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -322,6 +331,49 @@
 .attribute atomic_abi, 3
 
+
+#      UNKNOWN_UNKNOWN: BuildAttributes {
```
My other two comments were minor nits, this is the one where I'm wondering if there's some missing logic.

For every other case where one of the input files has the atomic_abi attribute and we produce an output, the output file also has the attribute. In this case, although one of the inputs had the attribute we end up dropping it in the output. I think that's probably not desired, but maybe I'm missing a reason to do this?

ilovepi replied:
So, we probably need to double check w/ @MaskRay, but I believe we don't emit attributes w/ 0 value in general. I think the following snippet seems to indicate that from a few lines below the call to `mergeAtomics`. I assume that is because the default value of 0 is pretty much 'not used', so it isn't emitted normally, and since UNKOWN is 0, it just falls into that case.

```
size_t size = 5 + merged.vendor.size() + 1 + 5;
  for (auto &attr : merged.intAttr)
    if (attr.second != 0)
      size += getULEB128Size(attr.first) + getULEB128Size(attr.second);
```

I'll update the patch to address the nits, but I'll try to wait for Fangrui to confirm if thats correct about 0 valued attributes before proceeding.

asb replied:
The main reason it _might_ be helpful to differentiate between the attribute being present but UNKNOWN and the attribute not being present at all, is that if the attribute is there (and emitted when merging UNKNOWN with NONE) you at least know it's been processed by tools that were aware of the attribute. Is that useful? Not sure. It's probably not hugely important either way, but thought I'd check if this was by design or not.

asb reviewed (commented):
One question in line about the observed output when combining a file without the ABI attribute with one that has it set to 'unknown', and two minor nits. If we resolve that question, it all looks good to me - thanks for your patience!

ilovepi committed:
update commit message

Created using spr 1.3.4

ilovepi committed:
Fix naming in test cases for consistency

Created using spr 1.3.4

ilovepi committed:
clang-format

Created using spr 1.3.4

ilovepi commented:
@MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.

asb reviewed (approved):
LGTM. Thanks!

lukel97 commented:
Hi, I'm seeing a segfault with GNU ld after this when running clang with `-march=rv64g` and riscv64-linux-gnu-ld 2.42. With LLD 18.1.3 I get a warning:
```
ld.lld: warning: /var/folders/yl/qbdtspsx7t97vqcl6dm6q9540000gn/T/--4b08f5.o:(.riscv.attributes): invalid tag 0xe at offset 0x45
```

I haven't tried it with a ToT lld yet but I thought I'd flag this in case something was accidentally made backwards incompatible!

asb commented:
Thanks Luke. I can reproduce locally (with GNU ld 2.42) so have gone ahead and reverted so we can decide what to do without breaking people's workflows in the meantime.

CC @kito-cheng 

It seems like there are two problems here, the first obviously by far the biggest:
* There seems to be a bug in GNU ld's elf attribute handling code that causes it to segfault when seeing an unrecognized attribute.
* Arguably warning whenever an unrecognized tag is encountered (as ld.lld does, via `ELFAttributeParser::parseAttributeList`) is overly verbose.

ilovepi commented:
@lukel97 Thanks for the report and @asb thanks for the reverts. 

It does look like we’ll need to wait for bfd to catch up or at least not crash.

MaskRay commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -1084,10 +1084,62 @@ static void mergeArch(RISCVISAInfo::OrderedExtensionMap &mergedExts,
+      it->getSecond() = AtomicABI::A6C;
+      return;
+    case AtomicABI::A7:
+      error(toString(oldSection) + " has atomic_abi=" + Twine(oldTag) +
```
This should use `errorOrWarn`.

Sorry that I did not follow up the discussion in time, but I'd appreciate that I had the opportunity to give this comment after the patch was approved by @asb ...

ilovepi replied:
Noted. I'll follow up in https://github.com/llvm/llvm-project/pull/90266. I based this on your comment to a similar line in https://github.com/llvm/llvm-project/pull/84598#discussion_r1521836857, but it seems like I misunderstood?

MaskRay commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -44,6 +44,39 @@
+# RUN: llvm-mc -filetype=obj -triple=riscv64  atomic_abi_A7.s -o atomic_abi_A7.o
+# RUN: not ld.lld atomic_abi_A6C.o atomic_abi_A7.o -o /dev/null 2>&1 | FileCheck %s --check-prefix=ATOMIC_ABI_ERROR --implicit-check-not=error:
+# ATOMIC_ABI_ERROR: error: atomic_abi_A6C.o:(.riscv.attributes) has atomic_abi=1 but atomic_abi_A7.o:(.riscv.attributes) has atomic_abi=3
+
```
We don't use double blank lines

MaskRay commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -44,6 +44,39 @@
+# ATOMIC_ABI_ERROR: error: atomic_abi_A6C.o:(.riscv.attributes) has atomic_abi=1 but atomic_abi_A7.o:(.riscv.attributes) has atomic_abi=3
+
+
+# RUN: llvm-mc -filetype=obj -triple=riscv64  atomic_abi_A6S.s -o atomic_abi_A6S.o
```
excess space

MaskRay commented on lld/test/ELF/riscv-attributes.s:
```diff
@@ -44,6 +44,39 @@
 # RUN: not ld.lld a.o b.o c.o diff_stack_align.o -o /dev/null 2>&1 | FileCheck %s --check-prefix=STACK_ALIGN --implicit-check-not=error:
 # STACK_ALIGN: error: diff_stack_align.o:(.riscv.attributes) has stack_align=32 but a.o:(.riscv.attributes) has stack_align=16
 
+## merging atomic_abi values for A6C and A7 lead to an error.
```
Mismatched

Upper case. And there is a better word than "merging"

MaskRay commented:
> @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.

I am sorry it took a long time for me to notice the discussion and make some comments about the linker code.
While I appreciate the LGTM, given the number of linker comments I had to give,  landing this 7 hours after LGTM feels a bit rushed.

ilovepi commented:
> > @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.
> 
> I am sorry it took a long time for me to notice the discussion and make some comments about the linker code. While I appreciate the LGTM, given the number of linker comments I had to give, landing this 7 hours after LGTM feels a bit rushed.

Apologies. That wasn't my intent. I'll be sure to allow some more time in the future.

asb commented:
> > @MaskRay @asb Are we happy w/ this patch? I think I've addressed most of the concerns raised thus far, but I want try and finish the attributes patches this week if possible, since I'm back from EuroLLVM + vacation.
> 
> I am sorry it took a long time for me to notice the discussion and make some comments about the linker code. While I appreciate the LGTM, given the number of linker comments I had to give, landing this 7 hours after LGTM feels a bit rushed. (If asb's LGTM was given on the day you commented, I'd certainly not oppose as it was my fault not noticing the change in 3 days.)

Sorry for that @MaskRay. I should have given an LGTM conditional on you or another lld dev explicitly lgtming the linker code. I think given jrtc27 had been involved in the thread and a discussion thread with you was mentioned I'd misjudged the level of attention this had got already from lld folks - so my bad.
